recipes/quickstart/agents/Agents_101/Tool_Calling_201.ipynb

```diff
@@ -0,0 +1,774 @@
+    "\n",
+    "- Step 5: For ease, we will extract first 80k words from the PDF and write these to a `.txt` file that we can summarise\n",
+    "\n",
+    "- Step 6: Use instances of `Meta-Llama-3.1-8b` instances to summaries the two PDF(s)\n",
+    "\n",
+    "- Step 7: Prompt the `70b` model to get the differences between the two papers being discussed"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 1: Defining the pieces\n",
+    "\n",
+    "We will start by describing all the modules from the image above, to make sure our logic works.\n",
+    "\n",
+    "In second half of the notebook, we will write a simple function to take care of the function calling logic"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Install necessary libraries"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 37,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!pip3 install groq\n",
+    "#!pip3 install arxiv\n",
+    "#!pip3 install tavily-python\n",
+    "#!pip3 install llama-toolchain\n",
+    "#!pip3 install PyPDF2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Necessary imports"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Note: PLEASE REPLACE API KEYS BELOW WITH YOUR REAL ONES"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 38,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os, arxiv, PyPDF2\n",
+    "from tavily import TavilyClient\n",
+    "from groq import Groq\n",
+    "\n",
+    "# Create the Groq client\n",
+    "client = Groq(api_key='gsk_PDfGP611i_HAHAHAHA_THIS_IS_NOT_MY_REAL_KEY_PLEASE_REPLACE')\n",
+    "\n",
+    "tavily_client = TavilyClient(api_key='fake_key_HAHAHAHA_THIS_IS_NOT_MY_REAL_KEY_PLEASE_REPLACE')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Main LLM thread: \n",
+    "\n",
+    "We will use a `MAIN_SYSTEM_PROMPT` and a `main_model_chat_history` to keep track of the discussion, since we are using 4 instances of LLM(s) along with this. \n",
+    "\n",
+    "Note, if you paid attention and notice that the SYSTEM_PROMPT here is different-thanks for reading closely! It's always a great idea to follow the official recommendations. \n",
+    "\n",
+    "However, when it's a matter of writing complex regex, we can bend the rules slightly :D\n",
+    "\n",
+    "Note, we will outline the functions here and define them as we go"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 50,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MAIN_SYSTEM_PROMPT = \"\"\"\n",
+    "Environment: iPython\n",
+    "Cutting Knowledge Date: December 2023\n",
+    "Today Date: 15 September 2024\n",
+    "\n",
+    "# Tool Instructions\n",
+    "- Always execute python code in messages that you share.\n",
```

Comment:
can you please elaborate a bit  on the difference of the prompt for activating custom tool calling between 101 notebook and here? trying to understand how should user think which way need to use?